src/cunumeric/set/unique.cu

```diff
@@ -125,6 +187,24 @@ static Piece<VAL> tree_reduce(Array& output,
                                  cudaMemcpyDeviceToDevice,
                                  stream));
       merged.destroy();
+
+      if (return_index) {
+        indices.first.destroy();
+        other_index.first.destroy();
+
+        auto buf_size_index =
+          (get_aligned_size(my_piece.second * sizeof(int64_t)) + sizeof(int64_t) - 1) /
+          sizeof(int64_t);
+        assert(my_piece.second <= buf_size_index);
+        indices.first = outputs[1].create_output_buffer<int64_t, 1>(buf_size_index);
+
+        CHECK_CUDA(cudaMemcpyAsync(indices.first.ptr(0),
+                                   merged_index.ptr(0),
+                                   sizeof(int64_t) * my_piece.second,
+                                   cudaMemcpyDeviceToDevice,
+                                   stream));
+        merged_index.destroy();
```

Comment:
I do not think it is safe for you to do a `cudaMemcpyAsync` and then destroy `merged_index` before the `cudaMemcpyAsync` is guaranteed to be done.

Reply:
Updating here it is presently safe to do this in Legate as all operations on a given GPU use the same stream. Since this is a tree reduction I think it's probably best to destroy without manually synchronizing to avoid excess latency and duplicating data across multiple iterations of the reduction.